### ISSUE TITLE
[Feature Request] DragType.FixedToPlane

### ISSUE THREAD

**Issue by terezo** [viromedia/virocore#107]
Hi,

I am trying to create ARRetail like android app. Currently I simply keeps one virtual plane Node attached to the biggest ARAnchor plane (supposed to be a room floor) which well extends its dimensions (eg. width and height 100f) in comparison to ARAnchor. The goal is to address drawbacks of plane recognition, since the corners of room are often not covered sufficiently with ARAnchor, what results with no scaling when user drag off an object from ARAnchor, as I can see in your ARPlacingObjects example.

I am setting onClick listener only on my virtual plane for placing an objects. I failed with adding objects as a child to clicked node (my virtual plane), because it keeps adding objects somewhere else. Also dragging those objects is weird, yet it has empty dragListener. Following is my displaScene() function with usage of addChildNode:

```
private fun ViroView.displayScene() =
        setScene(ARScene().apply {
            displayPointCloud(true)
            rootNode.addLight(AmbientLight(Color.WHITE.toLong(), 1000f))
            rootNode.addChildNode(createText(viroContext, vectorZero, "0, 0, 0"))

            setListener(ARSceneListener(this@ViroActivity) { _: Int, node: Node, clickPosition: Vector ->
                val modelPath = "file:///android_asset/object_coffee_mug.vrx"
                node.addChildNode(Draggable3DObject(modelPath).createOnPosition(node.convertWorldPositionToLocalSpace(clickPosition), this@ViroActivity))
//                rootNode.addChildNode(Draggable3DObject(modelPath).createOnPosition(clickPosition, this@ViroActivity)) // place correctly on clicked position
            })
        })
}
```

I guess it may have something to do with rotation of parent - it's created the same way as you ARAnchor plane tracking nodes. Here is my function:

```
fun createPlane(center: Vector, x: Float, z: Float, color: String = "#BF550000", type: Plane): Node {
        center.log("plane created witdth: $x, height: $z, type: ${type.name}")
        val plane = Surface(x, z)
        val material = Material()
        material.diffuseColor = Color.parseColor(color)
        plane.materials = Arrays.asList(material)
        val planeNode = Node()
        planeNode.geometry = plane
        planeNode.setRotation(Vector(-Math.toRadians(90.0), 0.0, 0.0))
        planeNode.setPosition(center)
        if (type == Plane.INFINITE) {
            planeNode.clickListener = object : ClickListener {
                override fun onClick(sourceID: Int, node: Node, vector: Vector) {
                    onClickPlane(sourceID, node, vector)
                }
                override fun onClickState(i: Int, node: Node, clickState: ClickState, vector: Vector) {}
            }
        }
        return planeNode
    }
```

So I keep adding it as a child node to root node for now, because it places objects correctly. 

My idea is to freeze `y` coordinate of objects placed on virtual plane so it sticks on floor, while it still scales at onDrag correctly at `x` and `z` coordinates. `DragType.FIXED_TO_NODE` sounds like exact behavior I need.

1. Could you please explain me how to use addChildNode function correctly? Do I need it in my case? 
2. Am I able to setPosition at onDragListener to keep `y` coordinates of my objects synced and stuck to  my virtual plane?
3. Why resulting Vector `node.convertWorldPositionToLocalSpace(clickPosition)` does not have `y` coordinate `0`, when clickPosition is an result of the click on the same `node` (my virtual plane)?

Thanks for your respond or pointing me out at any direction!
Roman

**Comment by terezo** [comment]
Hi @dthian ,

thanks for quick and exhausting response!

-> 1. Will try, but I am afraid it does not help me at this point, since # 2 (as I was expecting) dragging won't work :/

-> 2. I am playing for quite a while with ViroCore and know about both `DragType` values, but none of them seems to be sufficient for my cases. `DragType.FIXED_TO_WORLD` does not meet project requirements, cause we need to let users drag furniture into the corners, not only where the ARAnchor is being recognized and also do not want to jump over each recognized plane. That is unacceptable behavior.

-> 2.1. I believe new `DragType.FIXED_TO_NODE/PLANE/SURFACE` would have a lot more usages. For example, in still most popular AR furniture case, you do not want to throw vase from table or you want to push your chair under the table and see, if it fits, not drop it on the table.

Another one - imagine you are creating virtual city on your carpet and want to let kids drag the cars on the streets and parking lots, not crushing over the buildings.

Do you think trying to scale an object in drag listener `DragType.FIXED_DISTANCE` according to detected position (by `rayIntersectPlane`) on my virtual floor would work and worth the effort to achieve my goal?

Thanks a lot!

**Comment by dthian** [comment]
Hi @terezo - DragType.FIXED_TO_WORLD only drags objects to identified feature points found by ARCore in the real world. If there are no feature points found by ARCore (even though you see the corners of your room visually through the camera), ARCore doesn't know about feature locations / positions upon which we use to drag the object to. Thus, we only drag objects to locations we know for certain exists in the real world. **This is the intended behavior by design.** 

There is also merit for dragging objects on surfaces that adjusts dynamically to the world. Like dragging a vase across the floor, and then if your finger comes across a table surface, the vase is automatically dragged onto that surface. Like wise, a car on the streets can be dragged onto the top floor of a parking lot building that has an exposed top floor. 

However, as you mentioned, there is also merit to having a **DragType.FixedToPlane**, where users specifically want objects to be "locked" onto a plane when being dragged. For example when dragging a chair across the floor - to always drag on the XZ plane and be locked in a given Y plane. 

"Do you think trying to scale an object in drag listener DragType.FIXED_DISTANCE according to detected position (by rayIntersectPlane, smt like here: #35) on my virtual floor would work and worth the effort to achieve my goal?"
- By this do you mean being able to drag an object that is fixed to a plane, but is only draggable to the boundaries of the AR Plane? This is different from the proposed DragType.FixedToPlane above that just drags an object across an virtualized plane (that may be infinite in size). In this case, the user will have to scan his world until ARCore finds an ARPlane, and then, only then is this object dragged across the plane, within the boundaries of that plane, at the fixed depth of the plane - is this what you have in mind?

**Comment by terezo** [comment]
@dthian no, I want the proposed behavior. So DragType.FixedToPlane, which drags an object across an virtualized plane. Limit dragging to ARAnchor boundaries is exactly what I do not want to (to let user drag into room corners, even if not recoginzed as ARAnchor and keep scaling infinitely). I think FixedToNode would make more sense. Actually I am creating the virtualized plane only when ARAnchor is recognized. Basically I just want to extend the floor by infinite plane which has synced Y axis with the bigest ARAnchor, but do not want to limit dragging to ARAnchor.

**Comment by dthian** [comment]
@terezo, gotcha. Yes the filed feature request should cover the FixedToPlane scenario, where you will be able to drag you control on a virtualized plane. We are actively working on a fix and will update this issue once it has been pushed into release.

**Comment by terezo** [comment]
@dthian that would be amazing. I know how it is with estimations, but could you please provide me with rough estimate when it could be released? We have some deadlines and have to make some decisions. Anyway thanks a lot!

**Comment by dthian** [comment]
Hey @terezo, our next update with this feature request is scheduled to be deployed **roughly** around May 24th, plus minus a few days. Hope that helps!

**Comment by dthian** [comment]
This has been fixed in the latest release of ViroCore 1.7.2. 
Please checkout the update instructions here: https://virocore.viromedia.com/docs/releases

As mentioned in the docs:

> Dragging is limited to a user defined plane (specified by a point on the plane and its normal vector). You can also limit the maximum distance the dragged object is allowed to travel away from the camera/controller. These properties are controlled by setDragPlanePoint(Vector), setDragPlaneNormal(Vector), and setDragMaxDistance(float).

As such, both the planePoint and planeNormal actually define how the plane is oriented:

![40560549-5b1e3188-6028-11e8-9f84-625792b5882d](https://user-images.githubusercontent.com/1904934/41001776-cb17a1bc-68c6-11e8-97ea-bf68acd0d9ae.png)


Thus, if you want to drag the lock the object on a plane that is fixed to the Y Axis (the floor), you would:

node.setDragType(Node.DragType.FIXED_TO_PLANE)

// Sets the drag point in world coordinates
Vector planeOrigin = node.convertLocalPositionToWorldSpace(node.getPositionRealtime());
node.setDragPlanePoint(planeOrigin);         

// Sets the normal perpendicular to that point, creating a plane.  In this case, it's "Up": Y = 1
node.setDragPlaneNormal(0,1,0);                                

 // Sets the max distance you can drag an object along the plane.  
node.setDragPlaneMaxDistance(5);                            

You might wonder why the setDragPlaneDistance is necessary - this is because if you tilt your phone up (so that it is perpendicular to the floor), it is possible to drag an object along the floor into infinity. This would effectively push the object so far into the distance that it disappears :). To prevent this, we lock this draggable area with a maximum distance from the plane point.
